bitflow/pipeline.py

```diff
@@ -1,164 +1,149 @@
-import logging
 import multiprocessing
 import queue as thread_queue
-import threading
-import time
+from threading import Thread
 
-DEFAULT_QUEUE_MAXSIZE = 10000
-DEFAULT_MULTIPROCESSING_INPUT = True
-DEFAULT_BATCH_MULTIPROCESSING_INPUT = False
-DEFAULT_QUEUES_EMPTY_TIMEOUT = 0.1
+from bitflow.batchprocessingstep import *
+from bitflow.processingstep import ProcessingStep
 
 
-class Pipeline(threading.Thread):
+class PipelineTermination(ProcessingStep):
+
+    def __init__(self, sample_queue):
+        super().__init__()
+        self.__name__ = "Pipeline_Termination"
+        self.sample_queue = sample_queue
+
+    def execute(self, s):
+        super().write(s)
+        if s:
+            self.sample_queue.put(s)
+
+
+class Pipeline(Thread, metaclass=helper.CtrlMethodDecorator):
 
     def __init__(self, maxsize=DEFAULT_QUEUE_MAXSIZE, multiprocessing_input=DEFAULT_MULTIPROCESSING_INPUT):
+        super().__init__()
         self.__name__ = "Pipeline"
-        self.queue = None
+        self.lock = multiprocessing.Lock()
+        self.input_counter = multiprocessing.Value('i', 0)
         self.maxsize = maxsize
-        self.queue = self.create_queue(multiprocessing_input)
-        self.subqueues = {}
+        self.sample_queue_in = self.create_queue(multiprocessing_input)
+        self.sample_queue_out = self.create_queue(multiprocessing_input)
```

Comment:
Why is this pl needed?

Reply:
Pipelines can be within steps (like forks and batches) the queue_in is used to get samples into the parallel running pipe. The queue out is used to return results to the outer pipe for further processing. Its like:

Outer Pipe process ------------------------------------------------->
                       queue_in ----->|                        |<----queue_out
Inner pipe process                  --------------------